File: backend/app/schemas/project.py
from datetime import datetime
from pydantic import BaseModel, Field
from typing import Optional, List

from app.models.enums import ProjectStatus
from app.schemas.user import UserOut

class Address(BaseModel):
  city: Optional[str] = None
  street: Optional[str] = None
  house: Optional[str] = None
  apartment: Optional[str] = None

class ProjectBase(BaseModel):
  name: str
  description: str
  address: Optional[Address] = None
  budget: int
  team: Optional[List[int]] = Field(default_factory=list)
  start_date: datetime
  end_date: datetime
  status: ProjectStatus

class ProjectUpdate(BaseModel):
  name: Optional[str] = None
  description: Optional[str] = None
  budget: Optional[int] = None
  address: Optional[Address] = None
  team: Optional[List[int]] = None
  start_date: Optional[datetime] = None
  end_date: Optional[datetime] = None
  status: Optional[ProjectStatus] = None

class ProjectOut(ProjectBase):
  id: int
  author: UserOut
  spent_budget: int
  progress: float
  team: List[UserOut]
  created_at: datetime
  updated_at: datetime

  class Config:
    from_attributes = True

File: frontend/src/api/project.ts
import api from './client';
import type { User } from './user';

export type Address = {
  city?: string;
  street?: string;
  house?: string;
  apartment?: string;
};

export type ProjectType = {
  name: string;
  description: string;
  start_date: string;
  end_date: string;
  status: string;
  budget: number;
  spent_budget: number;
  team: User[];
  address: Address;
  author?: User;
  id?: number;
  progress: number;
  created_at?: string;
  updated_at?: string;
};

export type TaskType = {
  name: string;
  description: string;
  status: string;
  assigneed_to: number;
  priority: string;
  deadline: string;
  id: number;
  project_id: number;
  author: User;
  assignee: User;
  created_at: string;
  updated_at: string;
};

export type CreateProjectType = {
  name: string;
  description: string;
  start_date: Date;
  end_date: Date;
  status: string;
  budget: number;
  team: number[];
  address: Address;
};

export const getProjects = async () => {
  const response = await api.get('/projects/');
  return response.data;
};

export const createProject = async (projectData: CreateProjectType) => {
  const response = await api.post('/projects/', projectData);
  return response.data;
};

export const getOneProject = async (projectId: number) => {
  const response = await api.get(`/projects/${projectId}`);
  return response.data;
};

export const getTasks = async (projectId: number) => {
  const response = await api.get(`/projects/${projectId}/tasks`);
  return response.data;
};

export const updateProject = async (projectId: number, projectData: CreateProjectType) => {
  const response = await api.put(`/projects/${projectId}`, projectData);
  return response.data;
};
